### PR TITLE
Tudor/flakyness and perf fixes

### PR DESCRIPTION
**Pull request by tudor-malene**
### Why this change is needed

- The in-mem sim is flaky because there was a bug with chaining the batches when there is a reorg.
- Rollup decompression is slow on the sepolia testnet because of the inefficient block fetching. Which is now cached

### CODE REVIEW

File: go/enclave/components/rollup_consumer.go [hunk 1]
```diff
@@ -58,21 +58,34 @@ func (rc *rollupConsumerImpl) ProcessRollupsInBlock(b *common.BlockAndReceipts)
...
-				// todo - issue challenge as a validator
-				return err
-			}
-			if err := rc.storage.StoreRollup(rollup, internalHeader); err != nil {
-				rc.logger.Error("Failed storing rollup", log.RollupHashKey, rollup.Hash(), log.ErrKey, err)
-				return err
-			}
+
+	for _, rollup := range rollups {
+		l1CompressionBlock, err := rc.storage.FetchBlock(rollup.Header.CompressionL1Head)
```
Comment (tudor-malene) [inline go/enclave/components/rollup_consumer.go:63]: added some sanity checks here

File: integration/simulation/simulation_in_mem_test.go [hunk 2]
```diff
@@ -35,8 +35,8 @@ func TestInMemoryMonteCarloSimulation(t *testing.T) {
 		StartPort:                  integration.StartPortSimulationInMem,
 		IsInMem:                    true,
 		L1SetupData:                &params.L1SetupData{},
-		ReceiptTimeout:             30 * time.Second,
-		StoppingDelay:              10 * time.Second,
+		ReceiptTimeout:             5 * time.Second,
```
Comment (tudor-malene) [inline integration/simulation/simulation_in_mem_test.go:38]: those timeouts were too long for the in-mem sim

File: go/enclave/components/rollup_compression.go [hunk 3]
```diff
@@ -374,14 +406,27 @@ func (rc *RollupCompression) executeAndSaveIncompleteBatches(calldataRollupHeade
 		// check whether the batch is already stored in the database
 		b, err := rc.storage.FetchBatchBySeqNo(incompleteBatch.seqNo.Uint64())
 		if err == nil {
-			parentHash = b.Hash()
+			// chain to a parent only if the batch is not a reorg
+			if incompleteBatch.header == nil {
```
Comment (tudor-malene) [inline go/enclave/components/rollup_compression.go:410]: this is the bug that was causing the flakyness

File: go/enclave/components/rollup_compression.go [hunk 4]
```diff
@@ -374,14 +406,27 @@ func (rc *RollupCompression) executeAndSaveIncompleteBatches(calldataRollupHeade
...
+			}
 			continue
 		}
 		if !errors.Is(err, errutil.ErrNotFound) {
 			return err
 		}
 
 		switch {
+		// this batch was re-orged
+		case incompleteBatch.header != nil:
```
Comment (tudor-malene) [inline go/enclave/components/rollup_compression.go:421]: another rare flakyness issue, when the genesis itself was reorged

File: go/enclave/components/rollup_compression.go [hunk 5]
```diff
@@ -261,31 +263,30 @@ func (rc *RollupCompression) createIncompleteBatches(calldataRollupHeader *commo
...
+	}
+
+	l1Heights, err := rc.calculateL1HeightsFromDeltas(calldataRollupHeader, transactionsPerBatch)
+	if err != nil {
+		return nil, err
+	}
+
+	// a cache of the l1 blocks used by the current rollup, indexed by their height
+	l1BlocksAtHeight := make(map[uint64]*types.Block)
+	err = rc.calcL1AncestorsOfHeight(big.NewInt(int64(slices.Min(l1Heights))), rollupL1Block, l1BlocksAtHeight)
```
Comment (tudor-malene) [inline go/enclave/components/rollup_compression.go:279]: this is the performance fix to avoid traversing blocks

**Review by BedrockSquirrel — APPROVED** [review]
LGTM!

**Pull request merged**